### PR TITLE
Initial release.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Copyright (C) 2023 Toitware ApS.
+# Use of this source code is governed by a Zero-Clause BSD license that can
+# be found in the tests/TESTS_LICENSE file.
```
Comment (erikcorry) [inline Makefile]: Reference to a file that is not there.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,33 @@
-# toit-partition-table-esp32
-Toit package for handling ESP32 partition tables
+# ESP32 partition table parser
+
+A toit package and binary for handling ESP32 partition tables.
```
Comment (erikcorry) [inline README.md]: Don't we capitalize "Toit"?

File: bin/partitions.toit [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+import .src.esptool
+import .version
+
+main args:
+  // We don't want to advertise the '--version' on the root command
+  // as that would make it available to all sub commands.
+  // However, being able to write `app --version` is nice, so we handle
+  // that here.
+  if args.size == 1 and args[0] == "--version":
```
Comment (erikcorry) [inline bin/partitions.toit]: Our other tools also allow -v.

File: bin/partitions.toit [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    print_version
+
+  cmd := cli.Command "root"
+      --short_help="Commands to manage Ota partitions on the ESP32."
```
Comment (erikcorry) [inline bin/partitions.toit]:
```suggestion
      --short_help="Commands to manage OTA partitions on the ESP32."
```

File: bin/partitions.toit [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      --short_help="Print the version and exit."
+      --run=:: print_version
+
+  cmd.run args
+
+with_esptool parsed/cli.Parsed [block]:
+  esptool_path := parsed["esptool"]
+  port := parsed["port"]
+  partition_table_offset_str := parsed["partition-table-offset"]
+  partition_table_size_str := parsed["partition-table-size"]
```
Comment (erikcorry) [inline bin/partitions.toit]:
```suggestion
  partition_table_offset_str := parsed["partition-table-offset"].to-ascii-lower
  partition_table_size_str := parsed["partition-table-size"].to-ascii-lower
```

File: bin/partitions.toit [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    otadata := Otadata.decode otadata_bytes
+    2.repeat:
+      entry/SelectEntry := otadata.select_entries[it]
+      print """
+      otadata$it:
+        sequence-number: $entry.sequence_number
+        label: $entry.label.to_string_non_throwing
+        state: $entry.state ($(SelectEntry.state_stringify entry.state))
+        crc: $(%x entry.crc)"""
+      if it == 0: print
```
Comment (erikcorry) [inline bin/partitions.toit]: This `it` is a bit too far from its block start.

File: bin/partitions.toit [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    partition_table_bytes := esptool.read_partition_table
+    table := PartitionTable.decode partition_table_bytes
+    otadata_bytes := esptool.read_flash
+        --offset=table.find_otadata.offset
+        --size=table.find_otadata.size
+
+    otadata := Otadata.decode otadata_bytes
+
+    // For simplicity we always update entry 0 for ota_0 and entry 1 for ota_1.
+    // In theory
```
Comment (erikcorry) [inline bin/partitions.toit]: Missing 2nd half of comment?

File: bin/partitions.toit [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    }
+    new_state/int := state_mapping[parsed["state"]]
+    otadata.select_entries[index] = otadata.select_entries[index].with
+        --state = new_state
+        --sequence_number = sequence_number
+
+    print "new-state: $otadata"
+    new_otadata_bytes := otadata.encode
+
+    new_one := Otadata.decode new_otadata_bytes
```
Comment (erikcorry) [inline bin/partitions.toit]: Unused variable?

File: bin/src/esptool.toit [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+PARTITION_TABLE_OFFSET ::= 0x8000
+PARTITION_TABLE_SIZE ::= 0xc00
+
+class Esptool:
+  path_/string
+  port_/string?
+  partition_table_offset_/int
+  partition_table_size_/int
+
+
```
Comment (erikcorry) [inline bin/src/esptool.toit]:
```suggestion
```

File: bin/src/esptool.toit [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      return file.read_content out
+    unreachable
+
+  read_flash --offset/int --size/int --out/string:
+    args := [
+      "-b", "460800",
+      "read_flash",
+      "$offset",
+      "$size",
+      out,
```
Comment (erikcorry) [inline bin/src/esptool.toit]:
```suggestion
        "-b", "460800",
        "read_flash",
        "$offset",
        "$size",
        out,
```

File: bin/src/esptool.toit [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+      args += ["--port", port_]
+    run_ args
+
+  write_flash --offset/int --path/string:
+    args := [
+      "--after", "no_reset",
+      "-b", "460800",
+      "write_flash",
+      "$offset",
+      path,
```
Comment (erikcorry) [inline bin/src/esptool.toit]:
```suggestion
        "--after", "no_reset",
        "-b", "460800",
        "write_flash",
        "$offset",
        path,
```

File: bin/partitions.toit [hunk 12]
```diff
@@ -0,0 +1,255 @@
+// Copyright (C) 2023 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
```
Comment (erikcorry) [inline bin/partitions.toit]: Reference to a file that doesn't exist (here and other places).

File: src/otadata.toit [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  In the bootloader this state is changed to $STATE_IMAGE_PENDING_VERIFY
+  */
+  static STATE_IMAGE_NEW ::= 0
+  /**
+  First time this image has been booted.
+  The bootloader changes this state to $STATE_IMAGE_ABORTED.
+  */
+  static STATE_IMAGE_PENDING_VERIFY ::= 1
+  /**
+  The image has been marked as workable.
```
Comment (erikcorry) [inline src/otadata.toit]: "working"?  (Here and other places.)

File: src/otadata.toit [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  The partition can boot and work without limits.
+  */
+  static STATE_IMAGE_VALID ::= 2
+  /**
+  The image was neither marked as workable or non-workable.
```
Comment (erikcorry) [inline src/otadata.toit]:
```suggestion
  The image was neither marked as workable nor non-workable.
```

File: src/otadata.toit [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  crc/int
+
+  constructor --.sequence_number --.label --.state --.crc:
+
+  static decode bytes/ByteArray --offset/int -> SelectEntry:
+    ota_seq := LITTLE_ENDIAN.int32 bytes offset
+    offset += 4
+    seq_label_bytes := bytes[offset..offset + 20]
+    offset += 20
+    seq_label := seq_label_bytes.to_string_non_throwing
```
Comment (erikcorry) [inline src/otadata.toit]: Unused variable?

File: src/partition_table.toit [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    md5 := encode-md5-partition_ result[..cursor]
+    result.replace cursor md5
+    return result
+
+  encode-md5-partition_ partitions/ByteArray -> ByteArray:
+    checksum := md5.Md5
+    checksum.add partitions
+    partition := ByteArray 32: 0xff
+    partition.replace 0 MAGIC-BYTES-MD5
+    partition.replace 16 checksum.get
```
Comment (erikcorry) [inline src/partition_table.toit]:
```suggestion
    partition := ByteArray 32: 0xff
    partition.replace 0 MAGIC-BYTES-MD5
    partition.replace 16 (md5.md5 partitions)
```

**Review by erikcorry — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (floitsch) [reply Makefile]: done.

File: README.md (hunk 2 above)
Comment (floitsch) [reply README.md]: done.

File: bin/partitions.toit (hunk 12 above)
Comment (floitsch) [reply bin/partitions.toit]: done.

File: bin/partitions.toit (hunk 3 above)
Comment (floitsch) [reply bin/partitions.toit]: done.

File: bin/partitions.toit (hunk 4 above)
Comment (floitsch) [reply bin/partitions.toit]: done.

File: bin/partitions.toit (hunk 5 above)
Comment (floitsch) [reply bin/partitions.toit]: done. (slightly differently).

File: bin/partitions.toit (hunk 6 above)
Comment (floitsch) [reply bin/partitions.toit]: done.

File: bin/partitions.toit (hunk 7 above)
Comment (floitsch) [reply bin/partitions.toit]: removed the comment. Not even sure it made sense.
done.

File: bin/partitions.toit (hunk 8 above)
Comment (floitsch) [reply bin/partitions.toit]: done.

File: bin/src/esptool.toit (hunk 9 above)
Comment (floitsch) [reply bin/src/esptool.toit]: done.

File: bin/src/esptool.toit (hunk 10 above)
Comment (floitsch) [reply bin/src/esptool.toit]: we don't usually require list-entries to be intended by 4, but done.

File: bin/src/esptool.toit (hunk 11 above)
Comment (floitsch) [reply bin/src/esptool.toit]: we don't usually require list entries to be intended by 4, but done.

File: src/otadata.toit (hunk 13 above)
Comment (floitsch) [reply src/otadata.toit]: The docs seem to use "workable" from time to time. But they aren't native speakers either...
done.

File: src/otadata.toit (hunk 14 above)
Comment (floitsch) [reply src/otadata.toit]: done.

File: src/otadata.toit (hunk 15 above)
Comment (floitsch) [reply src/otadata.toit]: done.

File: src/partition_table.toit (hunk 16 above)
Comment (floitsch) [reply src/partition_table.toit]: done.

**Pull request merged**